Implement modified key export API for Mbed TLS 3.0

hanno-arm opened the pull request:
This resolves #4363 and implements the API proposed therein. See #4363 and the documentation for more information.

hanno-arm commented:
CI only failing because of the API break.

daverodgman commented on include/mbedtls/ssl.h:
```diff
@@ -976,6 +964,58 @@ struct mbedtls_ssl_session
+ * \param client_random The client random bytes.
+ * \param server_random The server random bytes.
+ * \param tls_prf_type The identifier for the PRF used in the handshake
+ *                     to which the key belongs.
```
Missing docs for secret and secret_len

hanno-arm replied:
Fixing

daverodgman commented on include/mbedtls/ssl.h:
```diff
@@ -976,6 +964,58 @@ struct mbedtls_ssl_session
+typedef enum
+{
+   MBEDTLS_SSL_TLS_PRF_NONE,
+   MBEDTLS_SSL_TLS_PRF_TLS1,
```
Is MBEDTLS_SSL_TLS_PRF_TLS1 still needed after TLS 1.0 and 1.1 have been removed?

hanno-arm replied:
Yes :+1:

hanno-arm replied:
@daverodgman I think we need to defer this because it breaks interoperability tests with OpenSSL which currently still rely on this constant.

daverodgman reviewed (changes requested):
A couple of minor points, otherwise looks good

hanno-arm commented:
@daverodgman Thank you for your review. I believe to have addressed your comments. Could you please re-review?

hanno-arm commented:
@ronald-cron-arm Why was this removed from the Mbed TLS 3.0 Optional EPIC?

ronald-cron-arm commented:
> 
> 
> @ronald-cron-arm Why was this removed from the Mbed TLS 3.0 Optional EPIC?

Because it is linked to the issue #4363 that is in the Mbed TLS 3.0 Optional EPIC and thus in the EPIC we can still see this PR through the issue.

hanno-arm commented:
@ronald-cron-arm Got it!

daverodgman reviewed (dismissed):
Looks good, comments have been addressed - thanks!

gilles-peskine-arm reviewed (changes requested):
I've read the discussion in https://github.com/ARMmbed/mbedtls/issues/2188 and https://github.com/ARMmbed/mbedtls/issues/4363 and reviewed the proposal here.

I think the callback should have access to the SSL context. This was proposed in https://github.com/ARMmbed/mbedtls/issues/2188 and I think this is both simpler and more future-proof than the proposed alternatives such as identifying a connection by its random. I don't see any rationale for the current choice of parameters for the callback.

hanno-arm commented on ChangeLog.d/key-export.txt:
```diff
@@ -0,0 +1,8 @@
+API changes
+   * mbedtls_ssl_conf_export_keys_ext_cb() has been removed.
```
Ok

hanno-arm replied:
Unless we're strict about this, I prefer keeping this under `API Changes`, because we've rather changed the export API than removed it.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
+ * \brief           Callback type: Export key alongside random values for
+ *                                 session identification, and PRF for
+ *                                 implementation of TLS key exporters.
+ *
```
I suggest to make the configuration SSL context specific instead, instead of passing the callback. This provides more flexibility to users.

hanno-arm replied:
I don't like to add getters for the key material because then you have to explain if/when is it save to call them, e.g. what happens if you call them when inside/outside a handshake, or a renegotiation, ... It's much easier to directly pass the key material when it's ready.

gilles-peskine-arm replied:
I don't think making the callback _function_ connection-specific is particularly useful. On the other hand, making the callback _context_ would be.

> you have to explain if/when is it save to call them

For the purpose here it's enough to guarantee that they're available during the call to the export callback.

hanno-arm replied:
Why add a global getter which must only be called from the key export callback, instead of adding a parameter to the callback? That doesn't make sense to me.

hanno-arm replied:
> I don't think making the callback function connection-specific is particularly useful. On the other hand, making the callback context would be.

I'd be OK with making the callback function itself configuration-specific, but the callback context connection-specific.

gilles-peskine-arm replied:
> Why add a global getter which must only be called from the key export callback, instead of adding a parameter to the callback?

Because if (when) we later discover that it's useful to access some other data in the callback, we can add a getter function. We can't add a parameter to the callback.

hanno-arm replied:
Note that the new design allows to add more elements to the enum. And if that happens not be enough (unlikely, IMO), we can still add getters which users can access using a reference to the SSL context which is part of the callback context.

hanno-arm replied:
I have made the callback and context connection-specific for now. I agree that the callback itself could be configuration specific, but I prefer to keep things together here.

gilles-peskine-arm replied:
I'm not convinced that it's better to pass all these arguments instead of letting the callback query them explicitly. But since the callback has access to the connection context, the API is extensible by adding getter functions, so the API is ok.

mpg replied:
After reviewing the discussions in #2188, #4363 and here, my gut feeling is it's unlikely but not impossible that the callback will need access to more information in the future, and I think the current API matches that nicely: it's optimized for the likely case when the callback needs what it currently does, by passing that as parameters (avoiding the need for getter functions), but also support extensions with getters should they be needed.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
+ *                     to which the key belongs.
+ *
+ * \return          \c 0 if successful.
+ * \return          A negative error code on failure.
```
Good point, I think the callback should return void indeed.

mpg replied:
I think returning `void` is the right thing to do for this kind of callback. If anything goes wrong with the export for some reason, it doesn't invalidate anything in the connection - it might be a problem for the application, and then the application will decide what to do (including closing the connection if desired).

daverodgman replied:
I think I agree with @mpg - not sure what Mbed TLS could usefully do with the error.

hanno-arm replied:
Changing to void.

hanno-arm replied:
For the record, I agree with @mpg's analysis.

hanno-arm commented:
@mpg @gilles-peskine-arm @daverodgman I've made the key export callback and context connection specific, addressing #2188. 

Could you please review?

gilles-peskine-arm commented:
You accidentally committed a modified `config.h`. Please amend the last commit to fix this.

gilles-peskine-arm commented on include/mbedtls/ssl.h:
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
+}
+mbedtls_tls_prf_types;
+
+#if defined(MBEDTLS_SSL_EXPORT_KEYS)
```
`MBEDTLS_SSL_EXPORT_KEYS` controls a very small amount of code, is not really a security risk (if you can corrupt the SSL context by adding a callback, you can surely corrupt the SSL context in many other ways), and has a wide variety of uses both for production and debugging. How about removing it (keeping it always on)? The fewer configuration options the better.

daverodgman replied:
I think this is probably a useful simplification, the impact of having it always-on looks extremely small.

hanno-arm replied:
I disagree here and would like to keep the option. Key export is for two use cases: (a) debugging, (b) application-specific key export. Both are non-standard, and are another potential source of key leakage - users should thus have the ability to disable them. (One might even contemplate disabling them by default.) Finally, while I haven't measured it, multiple invocations from multiple places of an unknown function pointer, with multiple arguments each, should be enough code that you'd like to be able to remove it on a highly constrained system.

Long story short: No change here for now.

gilles-peskine-arm replied:
> Both are non-standard, and are another potential source of key leakage - users should thus have the ability to disable them. (One might even contemplate disabling them by default.)

I don't see how that's relevant. The feature of invoking a callback is disabled by default, and even if you do register a callback, there is only a leak if the callback leaks the key.

> multiple invocations from multiple places of an unknown function pointer, with multiple arguments each

There's a single call site, that's not the issue. For an M0+ build of this PR at 95d69cf9471bd37cd5ea3553a17349e322eb69fc (baremetal config):
```
  350331  350027 (TOTALS)
   19971   19927 ssl_cli.o
   25364   25224 ssl_msg.o
   20787   20767 ssl_srv.o
   20246   20146 ssl_tls.o
```
For comparison, on `development` at 57864faf847572fa7b4e752068be55ce50bf2e8e:
```
  367608  367452 (TOTALS)
   20755   20751 ssl_srv.o
   20226   20102 ssl_tls.o
    4417    4389 version_features.o
```
I must say I'm surprised that the difference is so much larger in this PR — ~300B vs ~150B. I don't understand why there's an impact outside `ssl_tls.o`.

We're only testing the build with `MBEDTLS_SSL_EXPORT_KEYS` disabled in some reference configurations. So it's not untested, but it's slightly fragile (we may be missing some common combinations).

----

So now I would like to understand the code size impact, whether or not the compile-time option remains.

hanno-arm replied:
> I don't see how that's relevant. The feature of invoking a callback is disabled by default, and even if you do register a callback, there is only a leak if the callback leaks the key.

If we never make mistakes, you're right. But the mere fact that we're checking for a callback and passing a key to it increases our attack surface.

hanno-arm replied:
> So now I would like to understand the code size impact, whether or not the compile-time option remains.

I think we're making better use of our time by agreeing whether this feature can be disabled by default.

gilles-peskine-arm replied:
The default build is the one meant to be used when linking the library dynamically. So there should be a strong reason to disable a feature. The security risk is barely perceptible (it's in TLS which uses plenty of other function pointers anyway), so not a reason to disable in dynamic builds.

hanno-arm replied:
Fair point, if that's the established definition of the default config, I'm fine keeping it on by default.

It changes nothing w.r.t. my views to keep this configurable, though.

gilles-peskine-arm replied:
Explanation for the code size difference: it's due to short vs long field accesses in the context. After moving the export fields near the end of the context, I get:
```
   20222   20146 ssl_tls.o
```
So:

1. Please move the new fields towards the end of the context. It's a preexisting issue that the order of fields doesn't put often-used fields near the beginning, but let's not make it worse.
2. I still think we should remove `MBEDTLS_SSL_EXPORT_KEYS`. However, since this wasn't in the original scope of this pull request, and it can be done in 3.x since we'd just be making the feature always on, this is not a blocker here.

gilles-peskine-arm reviewed (changes requested):
Ok for the design of the callback. One day we should unify the design of the many callbacks, but today is not the day.

* I think we should remove `MBEDTLS_SSL_EXPORT_KEYS` as a configuration option.
* https://github.com/ARMmbed/mbedtls/pull/4552#discussion_r646175183 still needs adressing.
* The changelog entry could use some simplification.

daverodgman commented on ChangeLog.d/key-export.txt:
```diff
@@ -0,0 +1,13 @@
+     mbedtls_ssl_conf_export_keys_cb() have been removed
+     and replaced by a new API
+     mbedtls_ssl_set_export_keys_cb().
+   * The signature of key export callbacks configured via
+     mbedtls_ssl_set_export_keys_cb() is different from that
+     of the previous mbedtls_ssl_conf_export_keys_cb(): First,
```
Probably a fair comment, but IMO it is reasonable as it stands (users can ignore the second bullet if they aren't interested) - not worth investing lots of time on perfecting the changelog entry.

hanno-arm replied:
I'm OK swapping over to your wording.

daverodgman commented on include/mbedtls/ssl.h:
```diff
@@ -969,6 +957,60 @@ struct mbedtls_ssl_session
+ * \brief           Callback type: Export key alongside random values for
+ *                                 session identification, and PRF for
+ *                                 implementation of TLS key exporters.
+ *
```
If the answer might be "don't call it", should we document now that under some circumstances (that we don't need to define at this point), the callback may not be called?

hanno-arm commented:
>I think we should remove MBEDTLS_SSL_EXPORT_KEYS as a configuration option.

No change for now, see comment.

> #4552 (comment) still needs adressing.

Fixed.

> The changelog entry could use some simplification.

Fixed.

@gilles-peskine-arm @daverodgman @mpg Could you please re-review?

daverodgman reviewed (dismissed):
Looks good - I don't have a strong view on the MBEDTLS_SSL_EXPORT_KEYS question so I'm happy for this to land as-is (or with it removed if that ends up being the agreed approach).

hanno-arm commented:
@mpg @gilles-peskine-arm As explained in https://github.com/ARMmbed/mbedtls/pull/4552#discussion_r650030664, I'd suggest we take the opportunity of 3.0 to _disable_ this option by default. WDYT?

hanno-arm commented:
@gilles-peskine-arm argued that the default config is the one used for dynamic linkage. Users which obtain Mbed TLS as part of their distro should have the ability to make use of this feature, if only for debugging, thus we must keep it on by default. I withdraw my proposal to disable-by-default.

So there's no change pending from my side - I maintain the view that the feature should remain configurable.

mpg commented:
What's the status here? It looks to me like @hanno-arm has made the requested changes, but @gilles-peskine-arm labeled "needs: work" - is it only about the conflict that needs resolving, or do you also expect other changes related to previous feedback?

Also, I'm seeing two reviewers here already, so I'm removing myself as a reviewer; please protest if you think my review is needed here!

gilles-peskine-arm commented:
After further investigation from my last round of review, I [requested to add the export fields later in the context structure](https://github.com/ARMmbed/mbedtls/pull/4552#discussion_r650148282). I'm waiting for this to be done, and for CI to pass with no merge conflicts, before I re-review.

hanno-arm commented:
Moved the fields and rebased on top of `development` to remove numerous trivial conflicts related to the introduction of `MBEDTLS_PRIVATE()`.

hanno-arm commented:
@daverodgman @gilles-peskine-arm Can you please take another look?

daverodgman reviewed (approved):
This LGTM. The latest iteration changes an older commit (see git range-diff development 2369e3892 731870ad4 ) which seems to be fixing a mistake in the older version.

gilles-peskine-arm commented:
CI is still pretty unhappy.

gilles-peskine-arm commented:
CI's still unhappy.

daverodgman commented:
Rebased onto latest development to try re-running the CI

hanno-arm commented:
Yet another try...

daverodgman reviewed (approved):
LGTM

gilles-peskine-arm commented on docs/3.0-migration-guide.d/key-export.md:
```diff
@@ -0,0 +1,32 @@
+Those APIs have been removed and replaced by the new API
+`mbedtls_ssl_set_export_keys_cb()`. This API differs from
+the previous key export API in the following ways:
+
```
Another way is that the new callback returns void, whereas the old callbacks returned an int value that was ignored. Since the old value was ignored, that's minor enough not to be mentioned in the changelog which is a high-level semantic overview, but I think it should be listed in the migration guide, because I'd expect the list here to be a complete list of steps to migrate existing code.

daverodgman replied:
Fair point, I've discussed wording with @hanno-arm & pushed a commit with the addition to the migration guide.

gilles-peskine-arm reviewed (changes requested):
LGTM except for one omission in the migration guide.

daverodgman commented:
Sorry, typo in my first attempt.
